scripts/generate_tls_handshake_tests.py

```diff
@@ -0,0 +1,175 @@
+) -> None:
+    """Generate one TLS handshake defragmentation test.
+
+    :param out: file to write to.
+    :param side: which side is Mbed TLS.
+    :param length: fragment length, or None to not fragment.
+    :param version: protocol version, if forced.
+    """
+    #pylint: disable=chained-comparison,too-many-branches,too-many-statements
+
+    our_args = ''
+    their_args = ''
+
+    if length is None:
+        description = 'no fragmentation, for reference'
+    else:
+        description = 'len=' + str(length)
+    if version is not None:
+        description += ', TLS 1.' + str(version.value)
+    description = f'Handshake defragmentation on {side.name.lower()}: {description}'
+    tc = tls_test_case.TestCase(description)
+
+    if version == Version.TLS12 and \
+       length is not None and \
+       length >= TLS_HANDSHAKE_FRAGMENT_MIN_LENGTH and \
+       length < 16 and \
+       side == side.CLIENT:
+        # Skip test cases where the Finished message is fragmented in TLS 1.2.
+        # This is currently buggy when the symmetric encryption used an
+        # explicit IV (CBC, GCM or CCM; Chachapoly and null work, as does
+        # TLS 1.3, because they use a purely implicit IV).
+        tc.requirements.append('skip_next_test')
+
+    if version is not None:
+        their_args += ' -tls1_' + str(version.value)
+        # Emit a version requirement, because we're forcing the version via
+        # OpenSSL, not via Mbed TLS, and the automatic depdendencies in
+        # ssl-opt.sh only handle forcing the version via Mbed TLS.
+        tc.requirements.append('requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_' +
+                               str(version.value))
+        if side == Side.SERVER and version == Version.TLS12 and \
+           length is not None and \
+           length <= TLS12_CLIENT_HELLO_ASSUMED_MAX_LENGTH:
+            # Server-side ClientHello defragmentation is only supported in
+            # the TLS 1.3 message parser. When that parser sees an 1.2-only
+            # ClientHello, it forwards the reassembled record to the
+            # TLS 1.2 ClientHello parser so the ClientHello can be fragmented.
+            # When TLS 1.3 support is disabled in the server (at compile-time
+            # or at runtime), the TLS 1.2 ClientHello parser only sees
+            # the first fragment of the ClientHello.
+            tc.requirements.append('requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3')
+            tc.description += '  TLS 1.3 ClientHello -> 1.2 Handshake'
+
+    # To guarantee that the handhake messages are large enough and need to be
+    # split into fragments, the tests require certificate authentication.
+    # The party in control of the fragmentation operations is OpenSSL and
+    # will always use server5.crt (548 Bytes).
+    if length is not None and \
+       length >= TLS_HANDSHAKE_FRAGMENT_MIN_LENGTH:
+        tc.requirements.append('requires_certificate_authentication')
```

Comment:
I think if we move this out of the `if` we won't need the extra changes in the TLS PR that [I'm complaining about](https://github.com/Mbed-TLS/mbedtls/pull/10021#discussion_r1979233633).

Reply:
For the record: I think I was a victim of XYZ syndrome with the above comment: I want X (avoid changing OpenSSL default options in a way that impacts non-defragmentation tests), and instead I'm asking for Y (restore this arguably too-strong dependency) while perhaps there's a better way Z (add an explicit `-allow_no_dhe_kex` in the defragmentation tests, which would not only address X but also give us some defragmentation testing in PSK-only configs as a welcome bonus).

Reply:
I've changed to injecting `-allow_no_dhe_kex` where it's needed (or at least a superset of that which remains a subset of the handshake defragmentation test cases).